toolkit/tools/pkg/imagecustomizerlib/customizeuki.go

```diff
@@ -0,0 +1,455 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package imagecustomizerlib
+
+import (
+	"fmt"
+	"gopkg.in/ini.v1"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/microsoft/azurelinux/toolkit/tools/imagecustomizerapi"
+	"github.com/microsoft/azurelinux/toolkit/tools/imagegen/diskutils"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/file"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/logger"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/safechroot"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/safeloopback"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/safemount"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/shell"
+)
+
+const (
+	BootDir      = "boot"
+	UkiBuildDir  = "UkiBuildDir"
+	UkiOutputDir = "EFI/Linux"
+)
+
+func prepareUki(buildDir string, uki *imagecustomizerapi.Uki, imageChroot *safechroot.Chroot) error {
+	var err error
+
+	if uki == nil {
+		return nil
+	}
+
+	logger.Log.Infof("Enabling UKI")
+
+	// Check UKI dependency packages.
+	err = validateUkiDependencies(imageChroot)
+	if err != nil {
+		return fmt.Errorf("failed to validate package dependencies for uki:\n%w", err)
+	}
+
+	// Create necessary directories for UKI.
+	err = createUkiDirectories(buildDir, imageChroot)
+	if err != nil {
+		return fmt.Errorf("failed to create UKI directories:\n%w", err)
+	}
+
+	// This code installs the systemd-boot bootloader into the EFI system partition (ESP).
+	//
+	// The command being executed is:
+	//     bootctl install --no-variables
+	// This performs the following steps:
+	//   1. Creates the necessary directories in the ESP, such as:
+	//        - /boot/efi/EFI/systemd
+	//        - /boot/efi/loader
+	//        - /boot/efi/loader/entries
+	//   2. Copies the systemd bootloader binary from the host filesystem to the ESP:
+	//        - Copies /usr/lib/systemd/boot/efi/systemd-bootx64.efi to /boot/efi/EFI/systemd/systemd-bootx64.efi
+	//        - Copies /usr/lib/systemd/boot/efi/systemd-bootx64.efi to /boot/efi/EFI/BOOT/BOOTX64.EFI
+	//          (This second location serves as the fallback bootloader entry, adhering to UEFI conventions.)
+	//   3. Writes a random seed to /boot/efi/loader/random-seed. This is used by the bootloader to initialize randomness.
+	//
+	// The "--no-variables" flag ensures that the command does not modify UEFI NVRAM boot variables. Instead, it relies
+	// on the bootloader binaries being present in the ESP for booting.
+	err = imageChroot.UnsafeRun(func() error {
+		return shell.ExecuteLiveWithErr(1, "bootctl", "install", "--no-variables")
```

Comment:
When we support systemd-boot properly, this `bootctl install` call will likely be called by `hardResetBootLoader()`.